Migrate the TeamCity doc issue gen script to Go

nickvigilante opened the pull request:
Docs issues were created via a shell script. Now, they're created via a Go script. This will allow for testing and easier customization.

Fixes DEVINF-277 and CRDB-13554.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79861)
<!-- Reviewable:end -->

rickystewart commented on build/teamcity/cockroach/create-docs-issue.sh:
```diff
@@ -0,0 +1,16 @@
+source "$dir/teamcity-support.sh"
+source "$dir/teamcity-bazel-support.sh"
+
+BAZEL_SUPPORT_EXTRA_DOCKER_ARGS="-e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY -e TC_BUILD_BRANCH" run_bazel << 'EOF'
```
The list of environment variables in `BAZEL_SUPPORT_EXTRA_DOCKER_ARGS` should be exactly the set of environment variables your command will need to execute correctly. Doesn't look like you need any of these, but you do need `BUILD_VCS_NUMBER` and `GITHUB_API_TOKEN`.

nickvigilante replied:
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,155 @@
+	return result
+}
+
+func setMergeBranch(bs []byte) string {
```
Nit: I would expect this function to be called `getMergeBranch` rather than `setMergeBranch`. Same with `setCommits` below.

nickvigilante replied:
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,155 @@
+	} `json:"base"`
+}
+
+type ghPullCommit struct {
```
Some comments on these structs to capture what they are meant to represent would be good.

rickystewart replied:
Same with the functions, actually. For the best to include some comments if the name and signature doesn't tell the reader everything they need to know IMO.

nickvigilante replied:
Added some comments explaining what the different blocks do.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func prNums(bs []byte) []pr {
+	search := ghSearch{}
```
I think having all your functions take a `[]byte` argument is not what you want to do. Makes it very difficult to tell what each of these functions is meant to do, and I think is going to make your life much harder than you need when the time comes to test this. The caller can convert from `[]byte` to whatever type it needs (you can provide helper functions to do the same if you want). 

At the very least, if your function must take an unstructured array of bytes as an argument, a more informative parameter name than `bs` would help.

nickvigilante replied:
So `httpGet` returns a byte slice because it's the common ancestor among each of the 3 structs (ghSearch, ghPull, ghPullCommit) I created to house the data for each of my 3 helper functions. Within each helper function, I then unmarshal the byte slice into the struct of data.

I can create 3 more helper functions just to do the unmarshaling. Does that sound in line with your suggestion?

I also plan on extensively commenting the code, but only after I work through these nits, so that I don't accidentally document something that I have to change or remove.

rickystewart replied:
> So httpGet returns a byte slice because it's the common ancestor among each of the 3 structs (ghSearch, ghPull, ghPullCommit) I created to house the data for each of my 3 helper functions. Within each helper function, I then unmarshal the byte slice into the struct of data.

I get that's the intention, but it's still possible to achieve code reuse without passing around the raw byte slices to do all your work. You give up a lot of compile-time safety guarantees by doing that.

You can create more helper functions to do the unmarshaling. Reflection is also possible, like:


```
func httpGet(url string, token string, reflect.Type ty) (interface{}, error) {
	client := &http.Client{}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
      return nil, err
	}
	req.Header.Set("Authorization", "token "+token)
	res, err := client.Do(req)
	if err != nil {
      return nil, err
	}
	bs, err := io.ReadAll(res.Body)
	if err != nil {
      return nil, err
	}
	ret := reflect.New(ty)
  if err := json.Unmarshal(bs, ret); err != nil {
      return nil, err
  }
  return ret, nil
}

res, err := httpGet(url, token, (*ghSearch)(nil))
if err != nil {
    ....
}
search := res.(*ghSearch)
```

(We're not using generics yet so it's not an option.)

Not saying you have to use reflection, having specific helper functions for unmarshaling the 3 struct types you have might be easier, just demonstrating an alternate approach.

BTW: as this code block demonstrates, you can feel free to shadow the name `err`. No need to define many different `errN` variables.

nickvigilante replied:
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,155 @@
+	return pull.Base.Ref
+}
+
+func setCommits(bs []byte, p int) []commit {
```
As above, `setCommits` is an odd name for this. `getCommitsFromGhPullCommit` maybe? But I also don't really know what a `ghPullCommit` is supposed to represent  -- it's a commit, but also it captures many commits?

rickystewart replied:
Oh, also: what is `p`? Let's be careful of these one- or two-letter variable names unless they're so clear that further information is not necessary.

nickvigilante replied:
ghPullCommit is meant to be all commits within a single pull request in GitHub. I'm not sure what convention is on having plural type names, so I assumed singular was best. Got any suggestions about a better name? Happy to take them.

rickystewart replied:
Ah, yeah, see, I didn't get that this function takes an array of `ghPullCommit`s instead of a single `ghPullCommit`. I think `ghPullCommit` is as good a name as any.

rickystewart commented on pkg/cmd/docs-issue-generation/main.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+func maybeEnv(envKey, defaultValue string) string {
```
Consider using the utilities in `pkg/util/envutil` instead of re-defining this. `envutil.EnvOrDefaultString()`

nickvigilante replied:
Done.

Commit:
Migrate the TeamCity doc issue gen script to Go

Docs issues were created via a shell script. Now, they're
created via a Go script. This will allow for testing and
easier customization.

Release note: None

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+		_, err = client.Do(req)
+		//res.Body.Close()
+		if err != nil {
+			log.Fatalf("%s blablah2", err)
```
?

nickvigilante replied:
😆  Old debug code. Removed.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+	return result // return a slice of prs
+}
+
+func getMergeBranch(pull ghPull) string { // given a PR, return the base ref of that particular PR
```
Nit: Up to you, but I'm not sure this helper function is serving much of a purpose. It's only used once and it seems like it would be more readable to just inline it (`pull.Base.Ref` rather than `getMergeBranch(pull)`).

nickvigilante replied:
Deleted the function.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+	return nil
+}
+
+func prNums(search ghSearch) []pr { // given a GH search of PRs against a given commit, return an array of PRs
```
A unit test for this function wouldn't hurt.

nickvigilante replied:
This does require the GH API. Last time I tested this, it errored out, but I'll test it again and let you know how it goes.

nickvigilante replied:
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+	return pull.Base.Ref // once the data is unmarshaled, we search for the base ref of the PR in question
+}
+
+func getCommits(pullCommit []ghPullCommit, prNumber int) []commit { // given a specific PR, return a list of valid release commits on that PR
```
A unit test for this function also wouldn't hurt. The more complicated logic in this function is in `formatReleaseNote` which you already have tests for, so I don't expect a lot of duplication or anything, but a simple end-to-end test just to validate that this function is spitting out the correct output for a simple `[]ghPullCommit` input would be nice.

nickvigilante replied:
Will do.

nickvigilante replied:
Done.

rickystewart reviewed (changes requested):
Lots of comments but really most of the stuff I mentioned here is pretty small. We're very close to getting this merged :)

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+	re := regexp.MustCompile("(?s)[rR]elease [nN]ote \\(.*")                                       // this regex checks to make sure there's a release note within the commit
+	reNeg := regexp.MustCompile("([rR]elease [nN]ote \\(bug fix.*)|([rR]elease [nN]ote: [nN]one)") // this regex is used to exclude bug fixes or releases without release notes
+	rnSlice := re.FindAllString(message, -1)                                                       // return a slice of data that locates every instance of the phrase "release note (", case insensitive
+	if len(rnSlice) > 0 {                                                                          // checks to make sure a match was returned
```
Done.

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+	if len(rnSlice) > 0 {                                                                          // checks to make sure a match was returned
+		rn := rnSlice[0]                                // return the first instance of the desired release note
+		if len(rn) > 0 && !reNeg.MatchString(message) { // checks to make sure the desired release note is not null and doesn't match the negating string.
+			return "Related PR: https://github.com/cockroachdb/cockroach/pull/" + strconv.Itoa(prNumber) + "\nCommit: https://github.com/cockroachdb/cockroach/commit/" + sha + "\n\n---\n\n" + rn // add formatted release note
```
Done.

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+
+func formatTitle(message string) string {
+	reTitle := regexp.MustCompile("(?s)\n.*")
+	return reTitle.ReplaceAllString(message, "")
```
Done.

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,167 @@
+		req, _ := http.NewRequest("POST", postUrl, bytes.NewBuffer(reqBody))
+		req.Header.Set("Authorization", "token "+token)
+		_, err = client.Do(req)
+		//res.Body.Close()
```
Deleted.

nickvigilante commented on pkg/cmd/docs-issue-generation/main.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/spf13/cobra"
+)
+
+type parameters struct {
```
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,179 @@
+	PullRequest ghSearchItemPr `json:"pull_request"`
+}
+
+type ghSearchItemPr struct { // details about the pull request
```
Same comment as with functions: it is appropriate to put the comment describing the `struct` above the `struct` definition.

```
// ghSearchItemPr contains details about the pull request.
type ghSearchItemPr struct {
...
```

nickvigilante replied:
Done.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,179 @@
+			strconv.Itoa(prNumber),
+			sha,
+			rn,
+		) // return formatted release note
```
Comment seems not useful.

nickvigilante replied:
Deleted.

rickystewart commented on pkg/cmd/docs-issue-generation/docs_issue_generation_test.go:
```diff
@@ -0,0 +1,321 @@
+					{
+						Number: 80158,
+						PullRequest: ghSearchItemPr{
+							MergedAt: timeFromGhTime("2022-04-19T17:48:55Z"),
```
Nice!

nickvigilante replied:
Thank you 🥲

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,179 @@
+
+func httpGet(url string, token string, out interface{}) error {
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", url, nil) // initialize a new HTTP request
```
Done.

nickvigilante commented on pkg/cmd/docs-issue-generation/docs_issue_generation.go:
```diff
@@ -0,0 +1,179 @@
+	return nil
+}
+
+func prNums(search ghSearch) []pr { // given a GH search of PRs against a given commit, return an array of PRs
```
Done.

rickystewart commented:
Nice work!

bors r+
